app/services/push_notifications/event_notifier.rb

```diff
@@ -0,0 +1,24 @@
+module PushNotifications
+  class EventNotifier
+    def initialize(event:)
+      @event = event
+    end
+
+    # Who to notify
+    #
+    # @return [<User>]
+    def users
+      resource_logic.users
+    end
+
+    private
+
+    attr_accessor :event
+
+    def resource_logic
+      return PostLogic.new(event: event) if event.post_id
+
+      raise 'The resource associated to the Event is not supported'
+    end
+  end
```

Comment:
To me, this feels like a Factory. It knows what class to instantiate and but it also delegates `#users` to the underlying class. 

We can avoid this second unnecessary responsibility from `EventNotifier` by limiting it to return the `PostLogic` instance. This class is the one that actually implements `#users`. So something like:


```ruby
class Creator
  def create!
    event_notifier = EventNotifierFactory.notification_for(event: event)
    event_notifier.users.each do
      PushNotification.create!(event: event)
    end
  end
end
```

The idea here is that all kinds of `event_notifier` we come up with conform to the duck-type by implementing `#users`.

Reply:
Yeah, we need to play a bit more with the relationship between `PostLogic` and `EventNotifier`

Reply:
Look at `PostLogic` class as a way to organize things inside `EventNotifier` business logic. Users outside `EventNotifier` shouldn't know anything about it, they just ask `#users` to `EventNotifier`.